### PR TITLE
fix(codex): ensure [model_providers.<id>] TOML section exists in config.toml

### LINKED ISSUE #3449: CC-Switch 写入的 `config.toml` 格式不被 Codex CLI 识别，导致自定义路由配置始终不生效
### Self Checks / 自检

- [x] I have read the [FAQ](https://github.com/farion1231/cc-switch#faq) section in README.
我已阅读 README 中的[常见问题](https://github.com/farion1231/cc-switch#常见问题)。

- [x] I have searched for [existing issues](https://github.com/farion1231/cc-switch/issues), including closed ones.
我已搜索过[已有的 Issue](https://github.com/farion1231/cc-switch/issues)，包括已关闭的。

### CC Switch Version / 版本号

CC Switch v3.16.0

### Operating System / 操作系统

Windows

### Related App / 涉及应用

Codex

### Steps to Reproduce / 重现步骤

The user wants me to organize the issue they've been debugging so they can submit it to the CC-Switch author. Let me write a clear, well-structured bug report in Chinese.
### 问题标题

CC-Switch 写入的 `config.toml` 格式不被 Codex CLI 识别，导致自定义路由配置始终不生效

### 环境信息

- 操作系统：Windows 10 Home China (19045)
- Codex CLI 版本：0.135.0-alpha.1
- CC-Switch 版本：最新
- Codex 配置目录：`C:\Users\LENOVO\.codex\`

### 问题描述

CC-Switch 切换路由后写入的 `~/.codex/config.toml` 使用扁平格式，但 Codex CLI 要求自定义 provider 必须通过 `model_provider = "custom"` + `[model_providers.custom]` 段来声明。缺少此声明，Codex 默认使用 `openai` provider，顶层的 `base_url` 被完全忽略，所有请求直连 `api.openai.com`，导致 401 错误。

### CC-Switch 当前写入的格式（错误）

```toml
base_url = "http://127.0.0.1:15721/v1"
wire_api = "responses"
model = "astron-code-latest"
model_catalog_json = 'C:\Users\LENOVO\.codex\cc-switch-model-catalog.json'
experimental_bearer_token = "PROXY_MANAGED"
```

### Codex CLI 要求的格式（正确）

```toml
model_provider = "custom"
model = "astron-code-latest"
wire_api = "responses"
model_catalog_json = 'C:\Users\LENOVO\.codex\cc-switch-model-catalog.json'

[model_providers.custom]
name = "CC-Switch"
wire_api = "responses"
requires_openai_auth = false
base_url = "http://127.0.0.1:15721/v1"
experimental_bearer_token = "PROXY_MANAGED"
```

### 关键差异

| 项目 | CC-Switch 写入 | Codex 要求 |
|------|---------------|-----------|
| `model_provider` | ❌ 缺失 | ✅ 必须为 `"custom"` |
| `base_url` 位置 | ❌ 顶层，被忽略 | ✅ `[model_providers.custom]` 段内 |
| `requires_openai_auth` | ❌ 缺失 | ✅ 必须为 `false` |
| `[model_providers.custom]` 段 | ❌ 缺失 | ✅ 必须存在 |
| 实际行为 | 直连 api.openai.com → 401 | 走 127.0.0.1:15721 → CC-Switch |

### 验证方式

使用 `codex doctor` 命令可确认：

- **修复前**：`model provider: openai`，`endpoint: wss://api.openai.com/v1`，`auth mode: api_key`
- **修复后**：`model provider: custom`，`provider name: CC-Switch`，`custom API base URL: http://127.0.0.1:15721/v1`，`OpenAI auth not required`

### 复现步骤

1. 安装 CC-Switch 并切换路由到自定义 provider
2. 启动 Codex 桌面应用或 CLI
3. 所有请求直连 `api.openai.com`，CC-Switch 统计 0 命中
4. 运行 `codex doctor` 确认 `model provider` 仍为 `openai`

### 建议修复

CC-Switch 切换路由时，应将 `config.toml` 写为 Codex CLI 要求的 `[model_providers.custom]` 格式，而非扁平格式。核心改动：

1. 添加 `model_provider = "custom"` 到顶层
2. 将 `base_url`、`requires_openai_auth = false` 等字段移入 `[model_providers.custom]` 段
3. 保留其他已有配置段（如 `[marketplaces]`、`[plugins]`、`[windows]`、`[projects]`）不被覆盖

### Expected Behavior / 期望行为

CC-Switch 写入的 `config.toml` 格式不被 Codex CLI 识别，导致自定义路由配置始终不生效

### Actual Behavior / 实际行为

CC-Switch 写入的 `config.toml` 格式不被 Codex CLI 识别，导致自定义路由配置始终不生效

### Additional Context / 补充信息

_No response_

### PR DESCRIPTION
**Pull request by Kakueeen**
## Summary / 概述

Fix Codex CLI ignoring the `model_provider = "custom"` setting by ensuring the `[model_providers.<id>]` TOML section always exists in the written `config.toml`.

Codex CLI requires the `[model_providers.<id>]` section to route requests to a custom provider. When this section is missing — which can happen when Codex Desktop App rewrites `config.toml`, when CC-Switch imports a pre-existing flat-format config, or after a backup/restore cycle — Codex falls back to `model_provider: openai` and all API calls go directly to `api.openai.com`.

## Changes / 变更内容

| File | Change |
|------|--------|
| `src-tauri/src/codex_config.rs` | Add `ensure_codex_model_provider_section()` that defensively creates `[model_providers.<id>]` table when `model_provider` is set but the table is missing. Integrate into `write_codex_live_for_provider` and `prepare_codex_provider_live_config`. Add 6 unit tests covering normalization edge cases. |

### Fix details (commits 2-3)

1. **Non-implicit TOML table**: `toml_edit::Table::new()` creates implicit tables (no `[section]` header in output). Changed to `set_implicit(false)` so the synthesized `[model_providers.<id>]` header is emitted correctly.
2. **Test assertions**: Updated two test assertions to check `experimental_bearer_token` at its scoped location `[model_providers.<id>]` rather than the top level — the function intentionally scopes bearer tokens inside the provider table for custom providers.

## Root Cause / 根因

The `[model_providers.<provider>]` TOML table could be lost during the config lifecycle (Codex.app rewrites, import, backup restore). Without this table, Codex CLI ignores the `model_provider` top-level key and defaults to `openai`, routing all traffic to the official OpenAI endpoint instead of the configured custom provider.

## Test Plan / 测试计划

- [x] TypeScript type checking passes (`pnpm typecheck`)
- [x] `cargo test --lib codex_config` — 43 passed, 0 failed
- [x] 6 new unit tests added for `ensure_codex_model_provider_section`
- [ ] `cargo fmt --check` / `cargo clippy` (requires Rust 1.95)
- [ ] Manual verification: create a Codex third-party provider, verify `[model_providers.<id>]` section persists in `~/.codex/config.toml`

## Related Issue / 关联 Issue

Fixes #3449

### CODE REVIEW

**Comment by farion1231** [comment]
@codex review

File: src-tauri/src/codex_config.rs [hunk 1]
```diff
@@ -1037,6 +1037,63 @@ pub fn read_codex_live_settings() -> Result<Value, AppError> {
...
+    if doc.get("model_providers").is_none() {
+        doc["model_providers"] = toml_edit::table();
+    }
+
+    if let Some(mp) = doc.get_mut("model_providers").and_then(|item| item.as_table_mut())
+    {
+        if !mp.contains_key(&provider_id) {
+            let mut t = toml_edit::Table::new();
+            t.set_implicit(false);
+            mp.insert(&provider_id, toml_edit::Item::Table(t));
```
[the remark this answers is not in the document]
Comment (Kakueeen) [reply src-tauri/src/codex_config.rs]: Good catch — this is a valid concern. The flat config scenario (`model_provider = "vendor"` with top-level `base_url`/`wire_api`) can arise from CC-Switch importing a pre-existing flat-format config or from Codex Desktop App rewriting the config.

Pushed a fix in 6e29aa4 that makes `ensure_codex_model_provider_section()` migrate legacy top-level `base_url` and `wire_api` into the synthesized `[model_providers.<id>]` table when it's created, and clean up the top-level copies afterward.

Changes:
- `ensure_codex_model_provider_section()` now copies `base_url`/`wire_api` from top level into the new provider table (only when the table didn't previously exist — existing scoped fields are never overwritten)
- Added two unit tests: one covering the migration, another verifying that existing scoped values are preserved when the table already exists
- Also fixed `cargo fmt` issues that caused the CI backend check to fail

**Comment by Kakueeen** [comment]
@codex review

**Review by farion1231 — COMMENTED** [review]
感谢您的贡献。请查看一下是否有以下问题:

P1：Clippy 阻塞
cargo clippy --lib -- -D warnings 失败，位置在 src-tauri/src/codex_config.rs:1136：
.map(|t| ensure_codex_model_provider_section(t)) 应改成 .map(ensure_codex_model_provider_section)。

P1/P2：可能没覆盖 #3449 的字面复现
#3449 贴出的坏配置没有 model_provider，但 PR 的 ensure_codex_model_provider_section() 在缺少 model_provider 时直接原样返回。也就是说，它只修了“有 model_provider、缺 [model_providers.<id>] 表”的场景，可能没修“顶层 base_url/wire_api + 没有 model_provider”这个 issue 报告格式。需要作者确认真实现场格式，或补一条针对该 flat config 的 normalization。

P2：合成 provider 表时仍可能不完整
PR 会在没有任何 endpoint 字段时合成只有 experimental_bearer_token 的 [model_providers.<id>] 表，这是对旧测试 does_not_create_incomplete_provider_table 的反转。没有 base_url 的表对 Codex 路由帮助有限。另外，如果 provider 表已存在但缺 base_url/wire_api，而顶层有旧字段，当前代码会 early-return，不会补齐缺失字段。建议只在有可迁移 endpoint 时合成，或至少回填缺失字段并明确 name / requires_openai_auth 的策略。

File: src-tauri/src/codex_config.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kakueeen) [reply src-tauri/src/codex_config.rs]: Updated in a8d1bed to address all three review points:

**P1 (Clippy):** Changed `.map(|t| ensure_codex_model_provider_section(t))` to `.map(ensure_codex_model_provider_section)`.

**P1/P2 (Flat config without model_provider):** When the config has top-level `base_url` but no `model_provider` key (the exact format #3449 reported), the function now synthesizes `model_provider = "custom"` and creates `[model_providers.custom]` with the migrated endpoint fields. This covers the literal reproduction scenario from the issue where Codex Desktop App or other tools strip both `model_provider` and `[model_providers]`.

**P2 (Incomplete existing table):** When `[model_providers.<id>]` exists but is missing `base_url`/`wire_api` while top-level copies remain, the function now fills in the missing fields from the top level and cleans up the top-level copies. Existing scoped values are never overwritten.

Added 5 new tests covering:
- Flat config synthesis (no model_provider, has base_url)
- Noop when no model_provider and no endpoint fields
- Filling incomplete existing provider table
- Preserving existing scoped fields (not overwriting with stale top-level)
- Noop on empty config

**Comment by farion1231** [comment]
@codex review

File: src-tauri/src/codex_config.rs [hunk 2]
```diff
@@ -1037,6 +1037,124 @@ pub fn read_codex_live_settings() -> Result<Value, AppError> {
...
+                .get_mut(&provider_id)
+                .and_then(|item| item.as_table_mut())
+            {
+                for (field, val) in &fields_to_insert {
+                    provider_table.insert(field, val.clone());
+                }
+            }
+            // Clean up top-level copies after filling table (after provider_table borrow ends).
+            for (field, _) in &fields_to_insert {
+                doc.remove(field);
```
[the remark this answers is not in the document]
Comment (Kakueeen) [reply src-tauri/src/codex_config.rs]: Good catch — this is a real data loss risk.

**Problem:** When the provider entry is a TOML inline table (e.g. `[model_providers] vendor = { name = "Vendor" }`), `as_table_mut()` returns `None`. The previous code still executed `doc.remove()` for top-level `base_url`/`wire_api` after the failed insert attempt, silently dropping the endpoint fields that Codex CLI needs for routing.

**Fix in 2314748d:** Moved the cleanup loop inside the `if let Some(provider_table) = ...` block, so top-level fields are only removed after a successful insert into a genuine table. Inline table entries are left untouched with their top-level endpoint fields preserved.

**Test added:** `ensure_model_provider_section_preserves_flat_fields_for_inline_table` verifies that top-level `base_url` and `wire_api` are not removed when the provider entry is an inline table.
